### PR TITLE
Trimaran: Restructure for ease of adding trimaran plugins

### LINKED ISSUE #402: Trimaran: Restructure for ease of adding trimaran plugins
Currently, we have two load-aware trimaran plugins: TargetLoadPacking and LoadVariationRiskBalancing. As we plan to add more trimaran plugins, we need to extract common components for efficiency and ease of addition.

In particular, the following are common among trimaran plugins:

- Load watcher arguments
- A Collector to get data from load watcher, encapsulating the load watcher and its operations
- A PodAssignEventHandler to watch assigned pods and cache them locally
- A ResourceStats to analyze resource statistics data collected for a node

Above common functionality needs to be extracted into the trimaran package and have specifics to a particular trimaran plugin in its own sub package.

### PR DESCRIPTION
**Pull request by atantawi**
#### What this PR does / why we need it:
As we plan to add more trimaran plugins, we need to extract common components for efficiency and ease of addition.
#### Which issue(s) this PR fixes:

Fixes #402 
/kind cleanup

### CODE REVIEW

File: apis/config/types.go [hunk 1]
```diff
@@ -84,32 +84,36 @@ type MetricProviderSpec struct {
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
+// TrimaranArgs holds common arguments for trimaran plugins
+type TrimaranArgs struct {
+	metav1.TypeMeta
```
Comment (Huang-Wei) [inline apis/config/types.go:89]: I'd suggest embedding `MetricProvider` and `WatcherAddress` only.

**Review by Huang-Wei — COMMENTED** [review]
I will review the API parts. Leaving the trimaran specific refactoring (pkg/trimaran/*) to @wangchen615 and @zorro786

File: apis/config/types.go (hunk 1 above)
Comment (atantawi) [reply apis/config/types.go]: done

File: apis/config/scheme/scheme_test.go [hunk 2]
```diff
@@ -73,23 +73,23 @@ profiles:
         weight: 1
   - name: TargetLoadPacking
     args:
+      metricProvider: 
+        type: Prometheus
+        address: http://prometheus-k8s.monitoring.svc.cluster.local:9090
+      watcherAddress: http://deadbeef:2020
```
[the remark this answers is not in the document]
Comment (atantawi) [reply apis/config/scheme/scheme_test.go:79]: done

File: apis/config/v1beta2/types.go [hunk 3]
```diff
@@ -83,34 +83,40 @@ type MetricProviderSpec struct {
...
+	WatcherAddress *string `json:"watcherAddress,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:defaulter-gen=true
 
 // TargetLoadPackingArgs holds arguments used to configure TargetLoadPacking plugin.
 type TargetLoadPackingArgs struct {
-	metav1.TypeMeta `json:",inline"`
+	TrimaranArgs `json:",inline"`
```
[the remark this answers is not in the document]
Comment (atantawi) [reply apis/config/v1beta2/types.go]: done

[the remark this answers is not in the document]
Comment (atantawi) [reply apis/config/v1beta2/types.go]: Though, I had to do some manual conversions in apis/config/v1beta2/zz_generated.conversion.go

**Comment by atantawi** [comment]
/release-note-none
No visible changes, only internal naming and structuring.

File: apis/config/scheme/scheme_test.go [hunk 4]
```diff
@@ -73,23 +73,23 @@ profiles:
         weight: 1
   - name: TargetLoadPacking
     args:
+      metricProvider:
```
Comment (Huang-Wei) [inline apis/config/scheme/scheme_test.go:76]: nit: revert this as it's just reordering.

File: apis/config/scheme/scheme_test.go [hunk 5]
```diff
@@ -73,23 +73,23 @@ profiles:
...
-      metricProvider: 
-        type: Prometheus
-        address: http://prometheus-k8s.monitoring.svc.cluster.local:9090
   - name: LoadVariationRiskBalancing
     args:
       metricProvider:
         type: Prometheus
         address: http://prometheus-k8s.monitoring.svc.cluster.local:9090
         insecureSkipVerify: false
+      watcherAddress: http://deadbeef:2020
```
Comment (Huang-Wei) [inline apis/config/scheme/scheme_test.go:90]: ditto. revert this.

**Review by Huang-Wei — COMMENTED** [review]
some nits. LGTM otherwise.

BTW: `vendor/` changes is now required to be checked in. So could you run `go mod vendor && go mod tidy` to check in vendor changes **in a separate commit** if any?

**Pull request merged**